### PR TITLE
Add permission guard to midddleware

### PR DESCRIPTION
**Pull request by mr-bjerre**
Add new feature requested in #103.

Continued from #104

### CODE REVIEW

**Comment by boredland** [comment]
this looks really useful for me too! is there something blocking this @ferpintado and @silentworks ?

File: packages/nextjs/src/middleware/withMiddlewareAuth.ts [hunk 1]
```diff
@@ -1,15 +1,23 @@
...
   NextResponseMiddlewareAdapter,
-  jwtDecoder
+  jwtDecoder,
+  User
 } from '@supabase/auth-helpers-shared';
 
+class NoPermissionError extends Error {
+  constructor(message: string) {
+    super(message);
+    Object.setPrototypeOf(this, NoPermissionError.prototype);
```
Comment (silentworks) [inline packages/nextjs/src/middleware/withMiddlewareAuth.ts:17]: Is this necessary? it seems to be advised against in the [MDN](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/setPrototypeOf) docs.

File: packages/nextjs/src/middleware/withMiddlewareAuth.ts [hunk 2]
```diff
@@ -21,6 +29,10 @@ export interface withMiddlewareAuthOptions {
   redirectTo?: string;
   cookieOptions?: CookieOptions;
   tokenRefreshMargin?: number;
+  permissionGuard?: {
```
Comment (silentworks) [inline packages/nextjs/src/middleware/withMiddlewareAuth.ts:32]: I want to suggest we call this `authGuard` instead of `permissionGuard`.

**Comment by silentworks** [comment]
I've left two comments. Once we get these resolved I'll get this merged in/.

File: packages/nextjs/src/middleware/withMiddlewareAuth.ts (hunk 1 above)
Comment (boredland) [reply packages/nextjs/src/middleware/withMiddlewareAuth.ts:17]: ya, the typesafe and more readable variant would be with a simple override of the "name" getter. check it out in [this](https://www.typescriptlang.org/play?#code/MYGwhgzhAEByD2AFApgJwLYEsqfgOwFFVV5VpkAPAF2TwBMYiSyBvaAKGmmHwitQCuwKqQAU6ZFDABzZAC5ofVJjzSAlNBacuigQAc04yRBnI12gL7bZVaHjATRGtqmRUBqPNCoALbADoePCUhEVR-ewloKyt2UEgYAHlfNCZScmpaBmg01g4uIJDhMQkpWQUlFXVNbS4IfUNSk1lzLisuGzsHZCdNaFd3T28-CEDeflDSCO7o9lj+AE8argB6Fd8SAHc7ZG3c0QByMDx8BfR4ARg0ZgPW4a2d7YQUDGwIXEJiMQOT6ANXnD4chfVC3bQbeDbPC7aDJHypEGHeApMjXUhgizcMBUYA+aCia7ObSYABm+Ou0BUfGOwGQ8DJzzQWEBn2YRJ03F48BAyH8IHg0lEACIKfTvAsDNAhdAANTA8KRMyWcggCDISlkgnESnBKg0ulkuEItnLHSFbm8-mCkXasVUCXq6Vy67TCR3THIVXqrQc808vkC4UCPAAaxOmy8aLIdodUtl5O1kByILUrqVbTmQA) playground.

**Comment by boredland** [comment]
https://github.com/mr-bjerre/supabase-auth-helpers/pull/1

**Comment by silentworks** [comment]
@boredland I like how you created a PR for a PR 🤣. Nice one and thanks for reviewing this too.

**Comment by mr-bjerre** [comment]
Thanks for your feedback and the PR for the PR - thats great c,")

I've adapted the PR accordingly

File: packages/nextjs/src/middleware/withMiddlewareAuth.ts (hunk 1 above)
Comment (mr-bjerre) [reply packages/nextjs/src/middleware/withMiddlewareAuth.ts]: I'm getting this warning though. 

> Accessors are only available when targeting ECMAScript 5 and higher

File: packages/nextjs/src/middleware/withMiddlewareAuth.ts [hunk 3]
```diff
@@ -130,8 +130,8 @@ export const withMiddlewareAuth: withMiddlewareAuth =
       } else if (!authResult.user) {
         throw new Error('No auth user, redirecting');
       } else if (
-        options.authenticated &&
-        !(await options.authenticated(authResult.user))
+        options.authGuard &&
+        !(await options.authGuard.isPermitted(authResult.user))
```
Comment (boredland) [inline packages/nextjs/src/middleware/withMiddlewareAuth.ts:134]: uh, thanks. I wanted to do sth else initially and didn't revert it.

**Review by silentworks — APPROVED** [review]

**Comment by silentworks** [comment]
Thank you for this PR @mr-bjerre.

**Pull request merged**